src/frontend/Deprecation_analysis.ml

```diff
@@ -176,6 +184,20 @@ let rec collect_deprecated_stmt deprecated_userdefined
       acc
       @ collect_deprecated_lval deprecated_userdefined [] l
       @ collect_deprecated_expr deprecated_userdefined [] e
+  | Tilde {distribution= {name; id_loc}; arg; args; _}
+    when Option.is_some (String.Map.find deprecated_userdefined (name ^ "_log"))
+    ->
```

Comment:
One more `Option.(if is_some .. then value_exn ..)`. I also don't understand why this is here. It looks like this is detecting if the tilde statement refers to a user-defined `_log` density function. But the `_log` suffix is not visible here; in my opinion it would make more sense to warn about the function definition, not the places that use the function.

Reply:
We currently warn if a user writes 

`target += foo_log(x, ...)`
but not
`x ~ foo(...)` 

even though this are completely equivalent. Really it's the second one which is actually being deprecated/removed - if a user still wants to name a function `bar_log` in the future they can, it's just that it won't have the special meaning it does now. Warning at the definition time would suggest that they won't be able to do such a thing

Reply:
The warning at the definition suggest the meaning of the definition is going to change. Currently `_log` functions can use `_lupdf` and that's going to be deprecated. (which btw means that `target += foo_log()` is not _completely_ equivalent to a tilde)
I also think that use in tilde statement indicates a problem with the function definition which no longer matches the intended use.

Reply:
So are y0u suggesting we not warn at any usages, only at the definitions?

Reply:
I think
* Don't warn at calls. Those warnings are redundant if the definition emits a warning and unnecessary if the definition does not need to change.
* Don't warn at tildes. The tilde does not mention the deprecated suffix so even if the user takes action the statement itself will not change.
* The definition must warn if it either contains `_lupdf` or is used in a tilde, otherwise it's fine as is. That's kind of tricky to check, and the canonicalizer updates the name unconditionally so "always warn" is a reasonable choice here.

Reply:
I guess the tricky thing is that `_log` isn't technically a reserved suffix as far as I am aware, so the user should be able to write something like

```stan
functions {
   real add_one_log(real x){
     return log1p(x);
   }
}

model {
   real x = add_one_log(3.);
}
```

without it warning anywhere (though it looks like syntax highlighters are too broad on this as well), as this will be the sort of thing _log could totally be used for now and in the future. (as a historical note, it's not clear to me why both the _log suffix and the function multiply_log _both_ needed to be deprecated, but here we are)

If we want, I can try to make it warn only for Tildes and in the body of a definition where lup[dm]f is called? I agree that isn't the ideal location, but it should at least be doable and scoped how we want it to be.

Reply:
Yes, that's the problem. I'd rather just complain during the transition period and then allow it after `_log` distributions are completely gone. While your proposal is doable for the warnings module I don't see how to make it work in the canonicalizer.

Reply:
I've settled on warning at the definition, _only_ when:

1. The name ends in "_log"
2. The return type is either UReal or UInt

These are necessary conditions for it to be in the deprecation case, but sadly not sufficient to say it isn't a false alarm. I've updated the warning to account for this:

> ```
> Warning in 'log.stan', line 2, column 8: Use of the _log suffix in user
>     defined probability functions is deprecated and will be removed in Stan
>     2.32.0, use name 'add_one_log_lpdf' instead if you intend on using this
>     function in ~ statements or calling unnormalized probability functions
>     inside of it.
> ```

Canonicalize behavior also updated to make sure the return type is either real or int, otherwise it just leaves the name. Does this sound good?

Reply:
A function with `_log` suffix must have UReal return type, otherwise it does not pass typecheck
https://github.com/stan-dev/stanc3/blob/9802b07906eaa2014c59a1fdf7e5aa6747f247b4/src/frontend/Typechecker.ml#L1257-L1265

Reply:
Oh, what's all this business then? It seems like it's just wrong?

https://github.com/stan-dev/stanc3/blob/f1c30d28e143f4fe3fe790eb38c5bda872306c6a/src/frontend/Deprecation_analysis.ml#L75-L77

Reply:
That should not be the return type but the type of the first argument. Also I'm not sure if `is_real_type` is exactly correct there either, depends on if deeply nested arrays are allowed or not...

Reply:
```ocaml
let is_real_type = function
  | UReal | UVector | URowVector | UMatrix
   |UArray UReal
   |UArray UVector
   |UArray URowVector
   |UArray UMatrix ->
      true
  | _ -> false
```

besides my mixup, does that sound like an okay way to proceed?

Reply:
Oh, wow.
```stan
real foo_log(array[,] real x) {
  return sum(x[1]);
}
real foo_lpdf(array[,] real x) {
  return sum(x[1]);
}
```
The first one is fine, the second is apparently not? This is surely a bug but which one is wrong?

Reply:
Well that's certainly not ideal. I just checked and the same behavior is in 2.27, so it's not even something introduced by #995, it's just been around a while.

Reply:
I think it's because _log could be either a lpmf or lpdf, so it would be ambiguous whether it should be a real or int, so nothing is checked at all. I can trivially fix it in a separate PR to check that it's at least an int or a real but nothing else.

Reply:
I'm not sure if we want to be so restrictive. What's wrong with distribution of an array of arrays? `UnsizedType.is_real_type` and `is_int_type` could be changed to recurse deep into array types, that's how I'd expect them to behave based on their names anyway.

Reply:
I agree, as does @bob-carpenter. I've opened #1045 about it